CLI tools: write errors to stderr instead of stdout

williamrandolph opened the pull request:
Most of our CLI tools use the `Terminal` class, which previously did not provide methods for writing to standard output. When all output goes to standard out, there are two basic problems. First, errors and warnings are "swallowed" in pipelines, making it hard for a user to know when something's gone wrong. Second, errors and warnings are intermingled with legitimate output, making it difficult to pass the results of interactive scripts to other tools.

I considered adding a new `ERROR` verbosity level, which would, like `SILENT`, always be printed, but would be printed to standard error instead of standard output. But this didn't seem to fit the existing usages in, for example, `SetupPasswordTool`, where the `--verbose` flag controls whether or not stack traces are printed when errors happen. (If you think the "`ERROR` verbosity level" approach is better than the one I chose, please let me know.)

Instead, I added a second set of print commands to Terminal for printing to standard error, with `errorPrint` corresponding to `print` and `errorPrintln` corresponding to `println`. This leaves it to developers to decide which output should go where.

One little trick that I hope people will appreciate is that usage is printed to standard output when it's correctly requested (e.g., `bin/elasticsearch-keystore --help`) but goes to standard error when a command is invoked incorrectly (e.g. `bin/elasticsearch-keystore list-with-a-typo | sort`).

Fixes #43260

williamrandolph committed:
Add Standard-error writer to Terminal class

The java.io.Console class does not offer any support for writing to
Standard Error. Neither does the Terminal class we have written to wrap
usage of java.io.Console. This commit adds methods for writing messages
to standard error and partially addresses the first case described in
\#43260.

It's a first cut, and as such there's some probably unnecessary code
duplication in Terminal. I also still need to write unit tests.

williamrandolph committed:
Print usage to standard error for CLI errors

Writing usage and error messages to standard error is pretty, well,
standard for Unix utilities. It keeps errors and warnings from being
swallowed by pipelines to other utilities or being written to files.
This commit adds some flags to make sure that usage messages go to
stdout when requested with the '-h' or '--help' flags, but to stderr
when printed as a result of some program error.

I've also updated the MockTerminal so that we can check what's written
to stderr and added some TODOs for Terminal testing.

williamrandolph committed:
Send errors to stderr in various CLI utilities

This is my best effort to change our usage of the Terminal so that
errors and warnings are printed on Standard Error rather than Standard
Output. I'm not familiar with all of the tools I'm editing, so I will
need to do some further testing. If it turns out that I have made big
mistakes, I can revert just this commit and try again.

Relates #43260

williamrandolph committed:
Streamline error printing API

Users shouldn't have to figure the meaning of a boolean flag as an
argument to a print method. Instead, we'll use two public methods that
call a private method using the relevant boolean flag.

williamrandolph committed:
Checkstyle: add line breaks in very long lines

elasticmachine commented:
Pinging @elastic/es-core-infra

williamrandolph committed:
Fix Elasticsearch CLI tests

Instead of a Consumer<String> for output, use a BiConsumer<String,
String> for output plus error. This lets our tests be specific about
which output stream we expect messages to be written to.

williamrandolph committed:
Update plugin tests to look at stderr

williamrandolph committed:
Modify other tests to check stderr, not stdout

williamrandolph committed:
Fix tests for the SamlMetadata command

williamrandolph committed:
Fixing more tests to look at stderr

williamrandolph commented:
@elasticmachine Please update branch.

williamrandolph committed:
Merge branch 'master' into cli-errors-to-stderr

rjernst commented on distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/AddFileKeyStoreCommand.java:
```diff
@@ -55,8 +55,8 @@
         if (keystore == null) {
-            if (options.has(forceOption) == false &&
-                terminal.promptYesNo("The elasticsearch keystore does not exist. Do you want to create it?", false) == false) {
+            if (!options.has(forceOption) &&
```
We use `== false` instead of `!` because the former is much easier to visually notice when skimming through code. Please keep these as they were. :)

rjernst commented on server/src/test/java/org/elasticsearch/cli/TerminalTests.java:
```diff
@@ -41,6 +41,8 @@ public void testVerbosity() throws Exception {
         assertPrinted(terminal, Terminal.Verbosity.VERBOSE, "text");
     }
 
+    // TODO: test error verbosity
```
Can we add these new tests described here? I see several existing tests modified, but it would be good to have some direct coverage of the new functionality.

williamrandolph replied:
Of course - thanks for catching that.

rjernst reviewed (commented):
Looks good, 2 minor comments

williamrandolph committed:
Restore boolean comparison to correct style

Our team's code style prefers the "expression == false" to "!expression"
for readablity.

williamrandolph committed:
Add error output verbosity tests

I simply forgot my TODO items here. It's a good thing Ryan caught that
on code review, since it turns out I needed to add an extra call in the
MockTerminal for such tests to work correctly.

williamrandolph committed:
Merge branch 'master' into cli-errors-to-stderr
